Mathlib/Tactic/NormNum/GCD.lean

```diff
@@ -210,6 +210,53 @@ def evalIntLCM : NormNumExt where eval {u α} e := do
   let ⟨ed, pf⟩ := proveIntLCM ex ey
   return .isNat _ ed q(isInt_lcm $p $q $pf)
 
+
+theorem isInt_ratNum : ∀ {q : ℚ} {n : ℤ} {d : ℕ} {g : ℕ} {n' : ℤ},
+   IsRat q n d → Int.gcd n (Int.ofNat d) = g → IsInt (n / g) n' → IsInt q.num n'
+  | _, num, denom, g, _, ⟨hi, rfl⟩, hg, ⟨h'⟩ => by
+    subst g
+    have : 0 < denom := Nat.pos_iff_ne_zero.mpr <| by simpa using hi.ne_zero
+    constructor
+    simp_rw [Rat.mul_num, Rat.intCast_den, invOf_eq_inv,
+      Rat.inv_natCast_den_of_pos this, Rat.inv_natCast_num_of_pos this,
+      Rat.intCast_num, one_mul, mul_one]
+    rwa [Int.gcd, Int.ofNat_eq_natCast, Int.natAbs_ofNat] at h'
+
+theorem isNat_ratDen : ∀ {q : ℚ} {n : ℤ} {d : ℕ} {g : ℕ} {d' : ℕ},
+   IsRat q n d → Int.gcd n (Int.ofNat d) = g → IsNat (d / g) d' → IsNat q.den d'
+  | _, num, denom, g, _, ⟨hi, rfl⟩, hg, ⟨h'⟩ => by
+    subst g
+    have : 0 < denom := Nat.pos_iff_ne_zero.mpr <| by simpa using hi.ne_zero
+    constructor
+    simp_rw [Rat.mul_den, Rat.intCast_den, invOf_eq_inv,
+      Rat.inv_natCast_den_of_pos this, Rat.inv_natCast_num_of_pos this,
+      Rat.intCast_num, one_mul, mul_one, Nat.cast_id]
+    rwa [Int.gcd, Int.ofNat_eq_natCast, Int.natAbs_ofNat] at h'
+
+@[norm_num Rat.num _]
+def evalRatNum : NormNumExt where eval {u α} e := do
+  let .proj _ _ (q : Q(ℚ)) ← Meta.whnfR e | failure
+  have : u =QL 0 := ⟨⟩; have : $α =Q ℤ := ⟨⟩
+  have : $e =Q Rat.num $q := ⟨⟩
+  let _ : Q(DivisionRing ℚ) := q(inferInstance)
+  let ⟨q', n, d, eq⟩ ← deriveRat q
+  let eq : Q(IsRat $q $n $d) := eq
+  let ⟨gcd, pf⟩ := proveIntGCD q($n) q(Int.ofNat $d)
+  let ⟨lit, plit⟩ ← deriveInt q($n / $gcd) _
```

Comment:
I suspect this is suboptimal; either I should be calling some low-level procedure for computing division (or relying on it being `rfl`), or I should be skipping the low-level `proveIntGCD` and just using `deriveInt q($n / Int.gcd n $d)`.

@digama0, @kmill: do you have any preference?